django/db/models/fields/related_descriptors.py

```diff
@@ -1113,7 +1114,7 @@ def _add_items(self, source_field_name, target_field_name, *objs, through_defaul
             # source_field_name: the PK fieldname in join table for the source object
             # target_field_name: the PK fieldname in join table for the target object
             # *objs - objects to add. Either object instances, or primary keys of object instances.
-            through_defaults = through_defaults or {}
+            through_defaults = dict(resolve_callables(through_defaults or {}))
```

Comment:
I was about to suggest to move this line within the `if objs` branch but doing an early return on `if not objs` like we do in `_remove_items` is likely even better. That should likely be done in another commit.

Reply:
That's a good idea, thanks.

I'll do that change in a separate commit and rearrange the history so that it comes first.